### ISSUE TITLE
Pod readiness failure doesn't prevent ingress gateway pod from forwarding traffic

### ISSUE THREAD

**Issue by chenwng** [istio/istio#18374]
**Bug description**

I have a pod, which has readiness and liveness probes. For some reason, readiness probe fails and liveness probe still works. Pod is still shown as running but not ready and endpoint shows empty. Istio ingress gateway still forwards traffic to the pod.

It looks like pilot doesn't trigger any EDS push to ingress gateway pod if endpoint becomes empty.

Pilot log when pod readiness probe fails and endpoint becomes empty.
```
gc 877 @18630.107s 0%: 0.012+4.8+0.003 ms clock, 0.024+0.44/0.82/6.9+0.006 ms cpu, 11->11->5 MB, 12 MB goal, 2 P
2019-10-28T07:54:31.681417Z     info    Handling event update for pod cinfo-867fdbd745-lg6fx in namespace cinfo -> 192.168.1.28
2019-10-28T07:54:31.701158Z     info    Handle EDS endpoint cinfo in namespace cinfo -> 

```
**Expected behavior**

Expect ingress gateway pod stops forwarding traffic and only resumes forwarding after pod is ready again.

**Steps to reproduce the bug**
Here is an example pod I used to reproduce the issue.
1. Create a Istio cluster.
2. Create pod with readiness and liveness probe.  And pod is running normally. 
```
apiVersion: extensions/v1beta1
kind: PodSecurityPolicy
metadata:
  name: cinfo-psp
spec:
  seLinux:
    rule: RunAsAny
  runAsUser:
    rule: RunAsAny
  supplementalGroups:
    rule: RunAsAny
  fsGroup:
    rule: RunAsAny
  allowedCapabilities:
  - 'NET_ADMIN'
  volumes:
  - '*'
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: cinfo-role
  namespace: cinfo
rules:
- apiGroups: ['extensions']
  resources: ['podsecuritypolicies']
  verbs:     ['use']
  resourceNames: ['cinfo-psp']

---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: cinfo-rolebinding
  namespace: cinfo
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: cinfo-role
subjects:
- kind: ServiceAccount
  name: cinfo
  namespace: cinfo

---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: cinfo
  namespace: cinfo

---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: cinfo
  namespace: cinfo
  labels:
    app: cinfo
spec:
  selector:
    matchLabels:
      app: cinfo
  replicas: 1
  template:
    metadata:
      labels:
        app: cinfo
    spec:
      serviceAccount: cinfo
      containers:
      - name: cinfo
        image: nginx:alpine
        ports:
        - containerPort: 80
        livenessProbe:
          failureThreshold: 3
          initialDelaySeconds: 5
          periodSeconds: 5
          httpGet:
            path: /
            port: 80
            scheme: HTTP
        readinessProbe:
          failureThreshold: 1
          exec:
            command:
            - /bin/sh
            - -c
            - "[[ ! -e /fault ]] && exit 0"
---
apiVersion: v1
kind: Service
metadata:
  name: cinfo
  namespace: cinfo
spec:
  ports:
  - name: http-cinfo
    protocol: TCP
    port: 80
  selector:
    app: cinfo
```
3. Create Istio gateway/virtualservice. Test the pod can be accessed from external.

```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: cinfo-gateway
  namespace: cinfo
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http-cinfo
      protocol: HTTP
    hosts:
    - "*"
---
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: cinfo-virtual-service
  namespace: cinfo
spec:
  hosts:
  - "*"
  gateways:
  - cinfo-gateway
  http:
  - route:
    - destination:
        host: cinfo
        port:
          number: 80
    match:
    - port: 80

apiVersion: "authentication.istio.io/v1alpha1"
kind: "Policy"
metadata:
  name: cinfo-policy
  namespace: cinfo
spec:
  targets:
  - name: cinfo
  peers:
  - mtls:
      mode: PERMISSIVE

```
4. Make the pod readiness fail by execing into the pod and run touch /fault

5.  Check pod status and endpoint status. And test the pod accessibility again.

**Version (include the output of `istioctl version --remote` and `kubectl version`)**

istio 1.3.0

```
istioctl version
client version: c2bd59595ce699b31d0f931885f023028ff7902b
citadel version: c2bd59595ce699b31d0f931885f023028ff7902b
galley version: c2bd59595ce699b31d0f931885f023028ff7902b
ingressgateway version: c2bd59595ce699b31d0f931885f023028ff7902b
nodeagent version:
nodeagent version:
nodeagent version:
nodeagent version:
nodeagent version:
nodeagent version:
pilot version: c2bd59595ce699b31d0f931885f023028ff7902b
sidecar-injector version: c2bd59595ce699b31d0f931885f023028ff7902b
telemetry version: c2bd59595ce699b31d0f931885f023028ff7902b

kubectl version
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:13:54Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.3", GitCommit:"2d3c76f9091b6bec110a5e63777c332469e0cba2", GitTreeState:"clean", BuildDate:"2019-08-19T11:05:50Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}

```

**How was Istio installed?**

Installed with helm

**Environment where bug was observed (cloud vendor, OS, etc)**
```
cat /etc/redhat-release
Red Hat Enterprise Linux Server release 7.6 (Maipo)
```

**Comment by howardjohn** [comment]
This is a fixed bug but we need to port to the 1.3 branch https://github.com/istio/istio/pull/16967

**Comment by chenwng** [comment]
Thanks for the quick response! It seems cherry-picking failed due to conflict.

**Comment by howardjohn** [comment]
I will manually resolve it

**Comment by Mister101** [comment]
We have a similar issue with readinessProbes so figured I'd mention it here before making a new issue for it. In our case, I think the readinessProbe is succeeding when it shouldn't (as opposed to it failing and the ingress sending traffic to it anyway).

We have the `readiness.status.sidecar.istio.io/applicationPorts` annotation set to our container port, but the Deployment is still marked as ready before the proxy gets configured properly, giving us strange results where it returns a 503 on the first request, then is immediately fine on subsequent ones. Waiting approximately 2 minutes (in delay for readiness probe) seems to eliminate the first 503 as well.

**Comment by howardjohn** [comment]
@Mister101 we identified a few cases where the probe may be marked ready a few ms before traffic is accepted which we are working in reducing which may be what you are seeing. These are being worked on in master though so they aren't in any release yet and still need some testing.

This original issue has been fixed since 1.3.5
